k2/csrc/fsa.h

```diff
@@ -122,6 +126,139 @@ struct Fsa {
+// linearly as a CfsaVec with one element
+size_t GetCfsaVecSize(const Cfsa &fsa_in);
+
+/*
+  Create a CfsaVec from a vector of Cfsas (this involves representing
+  the vector of Fsas in one big linear memory region).
+
+     @param [in] fsas_in  The vector of Cfsas to be linearized;
+                      must be nonempty
+     @param [in] data    The allocated data of size `size` bytes
+     @param [in] size    The size of the memory block passed;
+                         must equal the return value of
+                         GetCfsaVecSize(fsas_in).
+ */
+void CreateCfsaVec(const std::vector<Cfsa> &fsas_in,
+                   void *data,
+                   size_t size);
+
+
+
+class CfsaVec {
+ public:
+
+  /*
+      Constructor from linear data, e.g. in a tensor holding a single
+      Fsa:
+         @param [in] size    size in int32_t elements of `data`, only
+                             needed for checking purposes.
+         @param [in] data    The underlying data.   Format of data is
+                             described below (all elements are of type
+                             int32_t unless stated otherwise).
+
+             - version       Format version number, currently always 1.
+             - num_fsas      The number of FSAs
+             - state_offsets_start  The offset from the start of `data` of
+                             where the `state_offsets` array is, in int32_t
+                             (4-byte) elements.
+             - arc_indexes_start   The offset from the start of `data` of
+                             where the `arc_indexes` array is, in int32_t
+                             (4-byte) elements.
+             - arcs_start    The offset from the start of `data` of where
+                             the first Arc is, in sizeof(Arc) multiples, i.e.
+                             Arc *arcs = ((Arc*)data) + arcs_start
+
+            [possibly some padding here]
+             - state_offsets[num_fsas + 1]   state_offsets[f] is the sum of
+                             the num-states of all the FSAs preceding f.  It is
+                             also the offset from the beginning of the
+                             `arc_indexes` array of where the part corresponding
+                             to FSA f starts.  The number of states in FSA f
+                             is given by state_offsets[f+1] - state_offsets[f].
+                             This is >= 0; it will be zero if the
+                             FSA f is empty, and >= 2 otherwise.
+            [possibly some padding here]
+
+             - arc_indexes[tot_states + 1]   This gives the indexes into the `arcs`
+                             array of where we can find the first of each state's arcs.
+
+             [pad as needed for memory-alignment purposes then...]
+
+             - arcs[tot_arcs]
+  */
+  CfsaVec(size_t size, void *data);
+
+  int32_t NumFsas() { return num_fsas; }
+
+  Cfsa &&operator [] (int32_t f) const;
```

Comment:
Why does it return an `rvalue reference`?

Reply:
I thought that was when you wanted to `move` the data?  So there aren't unnecessary constructors?  I may misunderstand.